applications/ShapeOptimizationApplication/custom_utilities/response_functions/face_angle_response_function_utility.cpp

```diff
@@ -0,0 +1,132 @@
+
+	mMainDirection = ResponseSettings["main_direction"].GetVector();
+	const double direction_norm = norm_2(mMainDirection);
+	KRATOS_ERROR_IF(domain_size != 3) << "FaceAngleResponseFunctionUtility: 'main_direction' vector norm is 0!" << std::endl;
+	mMainDirection /= direction_norm;
+
+	double min_angle = ResponseSettings["min_angle"].GetDouble();
+	mSinMinAngle = std::sin(min_angle * Globals::Pi / 180);
+
+	std::string gradient_mode = ResponseSettings["gradient_mode"].GetString();
+	if (gradient_mode == "finite_differencing")
+	{
+		double delta = ResponseSettings["step_size"].GetDouble();
+		mDelta = delta;
+	}
+	else
+		KRATOS_ERROR << "Specified gradient_mode '" << gradient_mode << "' not recognized. The only option is: finite_differencing" << std::endl;
+}
+
+double FaceAngleResponseFunctionUtility::CalculateValue()
+{
+	KRATOS_TRY;
+
+	const double value = block_for_each<SumReduction<double>>(mrModelPart.Conditions(), [&](Condition& rCond) {
+		const double g_i = CalculateConditionValue(rCond);
+		if (g_i <= 0) {
+			return 0.0;
+		}
+		return g_i*g_i;
+	});
+
+	mValue = sqrt(value);
+
+	return mValue;
+
+	KRATOS_CATCH("");
+}
+
+void FaceAngleResponseFunctionUtility::CalculateGradient()
+{
+	KRATOS_TRY;
+	// First gradients are initialized
+	VariableUtils().SetHistoricalVariableToZero(SHAPE_SENSITIVITY, mrModelPart.Nodes());
+
+	for (auto& cond_i : mrModelPart.Conditions()){
+
+		const double g_i = CalculateConditionValue(cond_i);
+		if (g_i <= 0) {
+			continue;
+		}
+
+		// Compute sensitivities using finite differencing in the three spatial direction
+		array_3d gradient(3, 0.0);
+
+		for (auto& node_i : cond_i.GetGeometry()){
+
+			// Apply pertubation in X-direction
+			double g_i_after_fd = 0.0;
+			node_i.X() += mDelta;
+			node_i.X0() += mDelta;
+			g_i_after_fd = CalculateConditionValue(cond_i);
+			gradient[0] = (g_i_after_fd - g_i) / mDelta;
+			node_i.X() -= mDelta;
+			node_i.X0() -= mDelta;
+
+			// Apply pertubation in Y-direction
+			g_i_after_fd = 0.0;
+			node_i.Y() += mDelta;
+			node_i.Y0() += mDelta;
+			g_i_after_fd = CalculateConditionValue(cond_i);
+			gradient[1] = (g_i_after_fd - g_i) / mDelta;
+			node_i.Y() -= mDelta;
+			node_i.Y0() -= mDelta;
+
+			// Apply pertubation in Z-direction
+			g_i_after_fd = 0.0;
+			node_i.Z() += mDelta;
+			node_i.Z0() += mDelta;
+			g_i_after_fd = CalculateConditionValue(cond_i);
+			gradient[2] = (g_i_after_fd - g_i) / mDelta;
+			node_i.Z() -= mDelta;
+			node_i.Z0() -= mDelta;
+
+			// Add to aggregated sensitivities
+			noalias(node_i.FastGetSolutionStepValue(SHAPE_SENSITIVITY)) += 1/mValue * g_i * gradient;
+		}
+	}
+
+	KRATOS_CATCH("");
+}
+
+double FaceAngleResponseFunctionUtility::CalculateConditionValue(Condition& rFace)
+{
+	// face normal
+	array_3d local_coords;
+	local_coords.clear();
+	array_3d face_normal = rFace.GetGeometry().Normal(local_coords);
```

Comment:
There is also `UnitNormal` function in the base geometry. you can use it here. 
Secondly, the 'local_coords' is empty, that is (0.0,0.0,0.0) here. is this intended ?

Reply:
> There is also UnitNormal function in the base geometry

:+1:

> the 'local_coords' is empty, that is (0.0,0.0,0.0) 

Yes, like this we compute the normal at the origin of the local coordinate system (center of a quad, corner point for a triangle)